File: 8_Dictionaries/javascript/DsDictionaryModule.js
export class KeyError extends Error {
  constructor(key) {
    super(`KeyError: ${String(key)}`);
    this.name = "KeyError";
    this.key = key;
  }
}

function reprPy(value) {
  if (typeof value === "string") {
    const escaped = value.replace(/\\/g, "\\\\").replace(/'/g, "\\'");
    return `'${escaped}'`;
  }
  if (Array.isArray(value) && value.length === 2) {
    return `(${reprPy(value[0])}, ${reprPy(value[1])})`;
  }
  return String(value);
}

function listStr(values) {
  return `[${values.map(reprPy).join(", ")}]`;
}

export class DsDictionaryView {
  constructor(store, viewType) {
    this._store = store; // reference to underlying store (live view)
    this._viewType = viewType;
  }

  toString() {
    if (this._viewType === "keys") {
      const keys = this._store.map(([k]) => k);
      return `DsDictionaryView_Keys(${listStr(keys)})`;
    }
    if (this._viewType === "values") {
      const values = this._store.map(([, v]) => v);
      return `DsDictionaryView_Values(${listStr(values)})`;
    }
    if (this._viewType === "items") {
      const items = this._store.map(([k, v]) => [k, v]);
      return `DsDictionaryView_Items(${listStr(items)})`;
    }
    throw new Error("Invalid view type");
  }

  *[Symbol.iterator]() {
    if (this._viewType === "keys") {
      for (const [k] of this._store) yield k;
      return;
    }
    if (this._viewType === "values") {
      for (const [, v] of this._store) yield v;
      return;
    }
    if (this._viewType === "items") {
      for (const [k, v] of this._store) yield [k, v];
      return;
    }
    throw new Error("Invalid view type");
  }
}

export class DsDictionary {
  constructor(items = null) {
    this._store = [];
    if (items != null) {
      for (const [k, v] of items) {
        this._store.push([k, v]);
      }
    }
  }

  toString() {
    const parts = this._store.map(([k, v]) => `${reprPy(k)}: ${reprPy(v)}`);
    return `{${parts.join(", ")}}`;
  }

  set(key, value) {
    for (let i = 0; i < this._store.length; i++) {
      const [k] = this._store[i];
      if (k === key) {
        this._store[i] = [key, value];
        return;
      }
    }
    this._store.push([key, value]);
  }

  get(key) {
    for (const [k, v] of this._store) {
      if (k === key) return v;
    }
    throw new KeyError(key);
  }

  delete(key) {
    for (let i = 0; i < this._store.length; i++) {
      const [k] = this._store[i];
      if (k === key) {
        this._store.splice(i, 1);
        return;
      }
    }
    throw new KeyError(key);
  }

  get length() {
    return this._store.length;
  }

  *[Symbol.iterator]() {
    for (const [k] of this._store) yield k;
  }

  keys() {
    return new DsDictionaryView(this._store, "keys");
  }

  values() {
    return new DsDictionaryView(this._store, "values");
  }

  items() {
    return new DsDictionaryView(this._store, "items");
  }
}

File: 8_Dictionaries/julia/DsDictionaryModule.jl
module DsDictionaryModule

export DsDictionary, DsDictionaryView, items

import Base: iterate, length, getindex, setindex!, delete!, show, keys, values

_py_repr(x) = string(x)

function _py_repr(x::AbstractString)
    # Python-ish repr for strings: single quotes
    escaped = replace(x, "\\" => "\\\\", "'" => "\\'")
    return "'" * escaped * "'"
end

function _py_repr(x::Tuple)
    # Only needed for 2-tuples in this unit
    if length(x) == 2
        return "(" * _py_repr(x[1]) * ", " * _py_repr(x[2]) * ")"
    end
    inner = join((_py_repr(v) for v in x), ", ")
    return "(" * inner * ")"
end

function _py_list_str(values)
    return "[" * join((_py_repr(v) for v in values), ", ") * "]"
end

mutable struct DsDictionary
    store::Vector{Tuple{Any,Any}}

    function DsDictionary(items=nothing)
        d = new(Tuple{Any,Any}[])
        if items !== nothing
            for item in items
                k, v = item
                push!(d.store, (k, v))
            end
        end
        return d
    end
end

struct DsDictionaryView
    store::Vector{Tuple{Any,Any}}
    view_type::Symbol
end

function show(io::IO, d::DsDictionary)
    parts = String[]
    for (k, v) in d.store
        push!(parts, _py_repr(k) * ": " * _py_repr(v))
    end
    print(io, "{" * join(parts, ", ") * "}")
end

function show(io::IO, v::DsDictionaryView)
    if v.view_type === :keys
        ks = [k for (k, _) in v.store]
        print(io, "DsDictionaryView_Keys(" * _py_list_str(ks) * ")")
        return
    end
    if v.view_type === :values
        vs = [val for (_, val) in v.store]
        print(io, "DsDictionaryView_Values(" * _py_list_str(vs) * ")")
        return
    end
    if v.view_type === :items
        its = [(k, val) for (k, val) in v.store]
        print(io, "DsDictionaryView_Items(" * _py_list_str(its) * ")")
        return
    end
    throw(ArgumentError("Invalid view type"))
end

function setindex!(d::DsDictionary, value, key)
    for i in eachindex(d.store)
        k, _ = d.store[i]
        if k == key
            d.store[i] = (key, value)
            return value
        end
    end
    push!(d.store, (key, value))
    return value
end

function getindex(d::DsDictionary, key)
    for (k, v) in d.store
        if k == key
            return v
        end
    end
    throw(KeyError(key))
end

function delete!(d::DsDictionary, key)
    for i in eachindex(d.store)
        k, _ = d.store[i]
        if k == key
            deleteat!(d.store, i)
            return d
        end
    end
    throw(KeyError(key))
end

length(d::DsDictionary) = length(d.store)

function iterate(d::DsDictionary, state::Int=1)
    state > length(d.store) && return nothing
    k, _ = d.store[state]
    return (k, state + 1)
end

function keys(d::DsDictionary)
    return DsDictionaryView(d.store, :keys)
end

function values(d::DsDictionary)
    return DsDictionaryView(d.store, :values)
end

function items(d::DsDictionary)
    return DsDictionaryView(d.store, :items)
end

function iterate(v::DsDictionaryView, state::Int=1)
    state > length(v.store) && return nothing
    k, val = v.store[state]
    if v.view_type === :keys
        return (k, state + 1)
    end
    if v.view_type === :values
        return (val, state + 1)
    end
    if v.view_type === :items
        return ((k, val), state + 1)
    end
    throw(ArgumentError("Invalid view type"))
end

end
